keymanapp/keyman
# feat(android): Feature/keyboard download manager

Fixes #1316

## Previous Behavior ##
Keyboard download in keyman android could take lot of time. Currently the user has to wait until download is finished because there is a modal dialog showing a infinite spinner.

## New Behavior ##
The Keyman keyboard is downloaded using the background DownloadManager using the same approach as the catalog download.
The Download happens in the background and the user can continue to use the application during the download. The user gets notified via a Android toast when the download is running and when it is finished.
The new keyboard is not automatically selected. The new Keyboard is marked with "[new]" in keyboard selection dialog.

Start Downloading new Keyboard from catalog
![Screenshot_1572505655](https://user-images.githubusercontent.com/52272980/67926627-061c9900-fbe9-11e9-803f-740cdb4587e8.png)

Confirm download
![Screenshot_1572505660](https://user-images.githubusercontent.com/52272980/67926644-16347880-fbe9-11e9-9fb6-778c2d889dab.png)

Download is started in background notification
![Screenshot_1572505663](https://user-images.githubusercontent.com/52272980/67926655-1f254a00-fbe9-11e9-84c0-0a16419a55cf.png)

Try to install the same keyboard again:
![Screenshot_1572505675](https://user-images.githubusercontent.com/52272980/67926690-3c5a1880-fbe9-11e9-8083-01d7a3774ed2.png)

Notification that download is finished
![Screenshot_1572505702](https://user-images.githubusercontent.com/52272980/67926711-4bd96180-fbe9-11e9-8662-48150f337d4b.png)

Additional feature: Mark don't switch keyboard when download is completed. Add new keyboard marker
![Screenshot_1572862339](https://user-images.githubusercontent.com/52272980/68113507-85320a00-ff26-11e9-85f5-cab3f7693583.png)

After selecting the keyboard the first time, new flag is removed.
![Screenshot_1572862354](https://user-images.githubusercontent.com/52272980/68113540-9a0e9d80-ff26-11e9-8478-5b392f230c5e.png)

## Work Done ##
- Start cloud download from KeyboardDownloadActivity
-  Add Cloud download Callbacks for Keyboardmetadata, keyboard package, fonts and dictionary
-  Flow chart for processes: This pull request contains processes concerning keyboard and dictionary/ lexical model update
[keyboard_download_diagram.pdf](https://github.com/keymanapp/keyman/files/3792080/keyboard_download_diagram.pdf)

## Next Steps ##
- Get meta data from keyboard cache
- Review notifications
- Remove old Downloadtask

## Testing 
- Goto Settings/ Installed languages
- Add a new language
- Select a keyboard from the list
- Download will start in background and send a Notification
- When download is completed, a Notification comes up, telling the resources are up to date.
- List of installed languages should show the new language
- Go back to the text view -- New Keyboard is not selected automatically
- Goto keyboard selection dialog
- the new keyboard is marked with "[new]"
- select the new keyboard
- reopen the keyboard selection dialog --> Keyboard is not marked as new
